src/modules/chat_tabcompletion/index.js

```diff
@@ -0,0 +1,146 @@
+const $ = require('jquery');
+const settings = require('../../settings');
+const watcher = require('../../watcher');
+const twitch = require('../../utils/twitch');
+const keyCodes = require('../../utils/keycodes');
+const emotes = require('../emotes');
+
+const CHAT_INPUT_SELECTOR = '.chat-input textarea';
+const AUTOCOMPLETE_SUGGESTIONS_SELECTOR = '[data-a-target=autocomplete-balloon]';
+
+function isSuggestionsShowing() {
+    return $(AUTOCOMPLETE_SUGGESTIONS_SELECTOR).length > 0;
+}
+
+const INPUT_EVENT = new Event('input', { bubbles: true });
+function setTextareaValue($inputField, msg) {
+    $inputField.val(msg)[0].dispatchEvent(INPUT_EVENT);
+}
+
+class ChatTabcompletionModule {
+    constructor() {
+        settings.add({
+            id: 'tabCompletionEmotePriority',
+            name: 'Tab Completion Emote Priority',
+            description: 'Prioritize emotes over usernames when using tab completion',
+            default: false,
+        });
+
+        watcher.on('load.chat', () => this.load());
+        watcher.on('chat.message', ($el, messageObj) => this.storeUser(messageObj.user));
+
+        $('body')
+            .on('keydown', CHAT_INPUT_SELECTOR, e => this.onKeydown(e))
+            .on('focus', CHAT_INPUT_SELECTOR, () => this.onFocus());
+    }
+
+    load() {
+        this.userList = new Set();
+        this.tabTries = -1;
+        this.suggestions = null;
+        this.textSplit = ['', '', ''];
+    }
+
+    storeUser(user) {
+        this.userList.add(user.userDisplayName || user.userLogin);
+    }
+
+    getSuggestions(prefix, includeUsers = true, includeEmotes = true) {
+        let userList = [];
+        let emoteList = [];
+
+        if (includeEmotes) {
+            emoteList = emotes.getEmotes()
+                .map(emote => emote.code)
+                .concat(this.getTwitchEmotes())
+                .filter(word => word.toLowerCase().indexOf(prefix.toLowerCase()) === 0);
+            emoteList.sort();
+        }
+
+        if (includeUsers) {
+            userList = this.getChatMembers().filter(word => word.toLowerCase().indexOf(prefix.toLowerCase()) === 0);
+            userList.sort();
+        }
+
+        if (settings.get('tabCompletionEmotePriority') === true) {
+            return [ ...emoteList, ...userList];
+        } else {
+            return [...userList, ...emoteList];
+        }
+    }
+
+    onFocus() {
+        this.tabTries = -1;
+    }
+
+    onKeydown(e, includeUsers = true) {
```

Comment:
rather than having two things bound to onKeydown of the textarea and duplicating a lot of code, you should have the chat history in here too.. they are related features, i dont get why you decoupled them when they were coupled for a reason